lib/ui/painting/codec.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/codec.h"
+
+#include "flutter/common/threads.h"
+#include "flutter/glue/trace_event.h"
+#include "lib/fxl/functional/make_copyable.h"
+#include "lib/tonic/dart_binding_macros.h"
+#include "lib/tonic/dart_library_natives.h"
+#include "lib/tonic/dart_state.h"
+#include "lib/tonic/logging/dart_invoke.h"
+#include "lib/tonic/typed_data/uint8_list.h"
+#include "third_party/skia/include/codec/SkCodec.h"
+
+using tonic::DartInvoke;
+using tonic::DartPersistentValue;
+using tonic::ToDart;
+
+namespace blink {
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, Codec);
+
+#define FOR_EACH_BINDING(V) \
+  V(Codec, framesCount)     \
+  V(Codec, repetitionCount) \
+  V(Codec, dispose)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+Codec::Codec(int frameCount, int repetitionCount)
+    : frameCount_(frameCount), repetitionCount_(repetitionCount) {}
+
+void Codec::dispose() {
+  ClearDartWrapper();
+}
+
+namespace {
+
+static constexpr const char* kInitCodecTraceTag = "InitCodec";
+
+std::unique_ptr<SkCodec> InitCodec(sk_sp<SkData> buffer, size_t trace_id) {
+  TRACE_FLOW_STEP("flutter", kInitCodecTraceTag, trace_id);
+  TRACE_EVENT0("blink", "InitCodec");
+
+  if (buffer == nullptr || buffer->isEmpty()) {
+    return nullptr;
+  }
+
+  return SkCodec::MakeFromData(buffer);
+}
+
+void InvokeCodecCallback(std::unique_ptr<SkCodec> codec,
+                         int frameCount,
+                         int repetitionCount,
+                         std::unique_ptr<DartPersistentValue> callback,
+                         size_t trace_id) {
+  tonic::DartState* dart_state = callback->dart_state().get();
+  if (!dart_state) {
+    TRACE_FLOW_END("flutter", kInitCodecTraceTag, trace_id);
+    return;
+  }
+  tonic::DartState::Scope scope(dart_state);
+  if (!codec) {
+    DartInvoke(callback->value(), {Dart_Null()});
+  } else {
+    fxl::RefPtr<Codec> resultCodec = Codec::Create(frameCount, repetitionCount);
```

Comment:
Why are the frameCount and repetitionCount extracted out of the SkCodec and passed separately?

Can the blink::Codec just take the SkCodec at construction time?  The blink::Codec constructor can then cache the frame count/repetition count if needed.

Reply:
The documentation for getFrameCount and getRepetitonCounts says it might cause the codec to read some more of the buffer data.
I figured that if we are already on the IO thread when we build the codec we should already get these counts...

I'm not sure it's really important the data comes from memory anyway...?

Reply:
Caching the counts probably makes sense.

But why does blink::Codec::Codec() receive the counts and then a separate set_codec() call provides the SkCodec?  Can the blink::Codec be constructed from the SkCodec?

Reply:
done